WIP: Feature/ompl rrt

marrts opened the pull request:
OMPL interface for Descartes, still a work in progress, needs some cleanup but working at this point.

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+template <typename FloatType>
+double descartes_light::DescartesStateSpace<FloatType>::getMaximumExtent() const
+{
+  return rung_to_rung_dist_ * static_cast<double>(ladder_rungs_.size() + 1);
```
Why `ladder_rungs_.size() + 1`? There are only `ladder_rungs_.size() - 1` rung connections

marrts replied:
Yes, `rung_to_rung_dist_ * static_cast<double>(ladder_rungs_.size() - 1)` would make the most sense. Initially I was both trying to account for index difference potentials and had a typo. There will be no index difference present in the maximum extent case because both the last and first rung only have 1 vertex since they are artificially added start and end states.

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+        double getMaximumExtent() const override;
+
+        /** \brief Returns the number of rungs in the ladder */
+        double getMeasure() const override;
```
What exactly does this mean?

marrts replied:
Looking into this further I realize that I have implemented this wrong. It is supposed to represent the "volume" of the space. So I'll have to multiply the number of rungs by the length of a rung.

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+  long unsigned int idx2 = state2->as<StateType>()->vertex.second;
+
+  // Initialize the distance between states at 0
+  double dist = 0;
```
```suggestion
  double dist = 0.0;
```

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+
+  // IF the states are equal then the distance is 0
+  if (rung1 == rung2 && idx1 == idx2)
+    return dist;
```
```suggestion
    return 0.0;
```

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+  // If they are on the same rung give a cost of 1 rung_to_rung_dist_, this should be seen as further away than points on adjacent
+  // rungs, but closer than points separated by 1 rung in between
+  if (rung_diff == 0)
+    dist += rung_to_rung_dist_;
```
See previous comment about whether this is conceptually possible

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+template <typename FloatType>
+bool descartes_light::DescartesStateSpace<FloatType>::equalStates(const ompl::base::State *state1, const ompl::base::State *state2) const
+{
+  long unsigned int rung1 = state1->as<StateType>()->vertex.first;
+  long unsigned int idx1 = state1->as<StateType>()->vertex.second;
+  long unsigned int rung2 = state2->as<StateType>()->vertex.first;
+  long unsigned int idx2 = state2->as<StateType>()->vertex.second;
+  if (rung1 == rung2 && idx1 == idx2)
+    return true;
+  return false;
```
I would define an equality operator for `StateType` instead

marrts replied:
I tried to do this and the operator definition didn't pass through when I tried to call `state1->as<StateType>() == state2->as<StateType>()`

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+    if (ladder_rungs_.size() == 0)
+        throw ompl::Exception("Ladder rungs cannot be empty");
+    if (ladder_rungs_.size() == 1)
+        throw ompl::Exception("Ladder rungs cannot be length 1");
```
```suggestion
    if (ladder_rungs_.size() < 2)
        throw ompl::Exception("At least 2 ladder rungs are required (" + std::to_string(ladder_rungs_.size()) + " provided)");
```

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/ompl_types.h:
```diff
@@ -0,0 +1,166 @@
+///*
```
Remove unused file

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/ompl_solver.h:
```diff
@@ -0,0 +1,72 @@
+ * algorithm with a default visitor to search the graph
+ */
+template <typename FloatType>
+class BGLOMPLSolver : public BGLSolverBaseSVDE<FloatType>
```
Per the previous comment, we shouldn't build this off of the BGL solver. It might make sense to inherit from the `LadderGraphSolver` though

marrts replied:
It now inherits directly from `Solver`

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/ompl_solver.h:
```diff
@@ -0,0 +1,72 @@
+
+  SearchResult<FloatType> ompl_search(std::shared_ptr<ompl::base::Planner> ompl_planner);
+
+  void initOMPL();
```
Do these need to be separate from the constructor and/or `search` method?

marrts replied:
I could combine `initOMPL()` with a redefinition of the constructor.

The `ompl_search` method is setup this way because I wanted the classes that inherit from it to easily call it without having to re-write everything in it. Also, it requires an input of a planner type, so it can't just inherit from `search`. This can't be just templated on `ompl::base::Planner` either because each planner has a unique set of parameters, so using the planner type as an input makes more sense to me.

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/ompl_solver.h:
```diff
@@ -0,0 +1,72 @@
+};
+
+using BGLOMPLRRTConnectSolverF = BGLOMPLRRTConnectSolver<float>;
+using BGLOMPLRRTConnectSolverD = BGLOMPLRRTConnectSolver<double>;
```
We should think about a way to make the specification of the search algorithm generic so we don't need to define new classes for every search algorithm we want to use. Depending on how exactly OMPL works, we could pass a search algorithm pointer into the constructor, or we could template this class on the type of search algorithm that should be used.

marrts replied:
One concern of mine is I'm not sure how many ompl algorithms will actually work with this setup, but I know RRT and RRTConnect do work. 

Your point still stands that it would be nice if one could just pass the ompl planner type. Unfortunately each planner has it's own unique set of parameters, so we would realistically need to account for every planner similar to how it's done in Tesseract. I think a better solution is to provide a method to `getSpaceInformation()` and `getMaxDistance()` from the base OMPLSolver class, which would allow a user to create the own instantiation of an ompl planner which then could be used to call `ompl_search`. It would look exactly like what is done in the `search()` method in `OMPLRRTSolver` and `OMPLRRTConnectSolver` but since it requires an input I think it should be kept separate from 'search()'

marip8 commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/ompl_solver.hpp:
```diff
@@ -0,0 +1,136 @@
+                                                                           max_dist_,
+                                                                           rung_to_rung_dist_);
+
+  dss_->setLongestValidSegmentFraction(0.5);
```
Should this be parameterized?

marrts replied:
So I don't think this actually does anything in the context of how I have the DescartesStateSpace methods defined, but it is required for OMPL. Basically it takes the maximum extent from the state space and multiplies by this factor. Any segment longer than this is checked at intervals equal to this length. We only ever check from point A to point B in Descartes, so I just set this to be a high value. (It also cannot be 1.0 or 0.0)

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+         * This value should be greater than any possible cost between adjacent rungs, however it should not
+         * be MAX_VALUE because the number of rungs away needs to be meaningful
+         * Example: The cost between rung 1 and rung 5 will be 4*rung_to_rung_dist_ */
+        const double rung_to_rung_dist_;
```
This and the `max_dist_` are the most complicated/confusing components of this whole search algorithm. In order to work with the OMPL algorithms we need some measure of distance between any two arbitrary states, so we need to somehow represent that, for [rung,index], [1,2] is closer to [3,4] than [50,2]. This is where `rung_to_rung_dist_` comes into play. To me, it seems that it should also be true that [1,2] is closer to all [2,x] than all [3,y] (I'm actually not certain that this is critical). Therefore, the `rung_to_rung_dist_` should be greater than any adjacent connection. At this moment 10000 is totally arbitrary, but I didn't know what a meaningful default value would be. Theoretically the user should define a meaningful cost given their specific problem.

This same logic applies to the `max_dist_` parameter which should be set to work for each specific problem. `max_dist_` represents the maximum allowed cost for any connection to be valid and that will vary dramatically given each application. I envisioned that likely a user would start with a high `max_dist_` and then see what solution is given and perform further tuning/refinement on the value until more optimal solutions are given.

Given everything I said above, especially the statement of 

> To me, it seems that it should also be true that [1,2] is closer to all [2,x] than all [3,y] (I'm actually not certain that this is critical).

I think maybe it would make sense to make `rung_to_rung_dist_` something like `1000*max_dist_` rather than a uniquely defined value. I believe this should still demonstrate the same behavior of converging on a solution.

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+        const double rung_to_rung_dist_;
+        /** @brief This is the minimum assigned cost for when a edge is evaulated to have no cost
+         * which is invalid for OMPL algorithms */
+        const double distance_epsilon_ = 0.000001;
```
Good point, this makes much more sense

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+        /** @brief Edge evaluator used to evaluate the cost of edges */
+        const std::vector<typename descartes_light::EdgeEvaluator<FloatType>::ConstPtr> edge_eval_;
+        /** @brief Maximum allowed cost for an edge to be valid */
+        const double max_dist_;
```
See reply about `rung_to_rung_dist_`

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+    public:
+
+        /** \brief The definition of a state in the Descartes State Space */
+        class StateType : public ompl::base::State
```
OMPL documentation specifically dictates that this should always be named `StateType`, otherwise I'd agree with you.

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+         * If they are more than 1 rung apart it will have a base distance of rung_to_rung_dist_*(rung difference) + (index difference).
+         * If they are 1 rung apart it will evaluate the cost of the edge.
+         * The rung it is going towards will then have the cost of that vertex
+         * added unless it is going to the goal vertex which has a cost of 0. */
```
> Is there an implicit requirement here that the states in each rung be "sorted" such that the same index in each rung represents a similar state (in terms of configuration, etc.)? If so, we should explicitly call out this assumption

The assumption is that these states are not sorted in any way.

> Why? Shouldn't the distance between states on the same rung be infinite because it not conceptually possible to visit multiple states on the same rung

I don't think it actually matters what the distance on the same rung is. As long as it's greater than `max_dist_` and triggers an interpolation. I wanted to set it to 1 rung apart because in my way of thinking while it is impossible to connect [1,2] ([rung,index]) to [1,4], it is also impossible to connect [1,2] to [50,2], but when visualizing those three points in the graph, [1,2] is geometrically much closer to [1,4] than [50,2]. However you are right that [1,2] and [50,2] could both exist in a solution whereas [1,2] and [1,4] could not. In the end I believe both will result in the exact same behavior given how I have the interpolation algorithm set up. The distance just has to be less than the value returned from `getMaximumExtent`.

> I don't understand what you're trying to say here. Please clarify

I will more explicitly explain everything in the description. What I meant here is that the total cost for a connection from point A to point B is the cost of the edge plus the cost of point B. Except for the case where point B is the goal state, which should inherently have a cost of 0, so it is not evaluated.

marrts commented:
I'm going to go ahead and convert this so that it doesn't have any boost dependencies, then I'll rebase off the re-organization that was recently merged. From there we can tackle some of your specific feedback points more directly.

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+          , rung_to_rung_dist_(rung_to_rung_dist)
+          , stateBytes_(sizeof(std::pair<long unsigned int, long unsigned int>))
+        {
+            type_ = 14; // Larger than default types
```
Yes, ompl provides `ompl::base::STATE_SPACE_TYPE_COUNT`, I will change to that

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+        /** \brief Boost graph ladder, used to help with edge and vertex cost evaluation */
+        descartes_light::BGLGraph<FloatType> graph_;
+        /** @brief Ladder graph representation of the graph vertices, used for creating edge connections */
+        std::vector<std::vector<VertexDesc<FloatType>>> ladder_rungs_;
```
No more boost deps

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+        {
+        public:
+            /** \brief The definition of the vertex (rung, idx) */
+          std::pair<long unsigned int, long unsigned int> vertex;
```
I agree, the rung and idx makes more sense and matches what is seen elsewhere in Descartes. Not sure what motivated me to go with the pair in the first place.

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+        std::size_t stateBytes_;
+    };
+
+    /** \brief A motion validator for the Descartes State Space */
```
Changed to `Ensures only motions that increment by +1 rungs are valid`

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/descartes_space.h:
```diff
@@ -0,0 +1,204 @@
+    };
+
+    /** \brief A state space representing the Descartes boost graph. The distance function is determined by the difference
+     * in rungs and indices unless the rungs are adjacent, then an actual cost is made calculated using an edge evaluator */
```
Added some more details, let me know what you think

marrts commented on descartes_light/solvers/include/descartes_light/solvers/ompl/impl/descartes_space.hpp:
```diff
@@ -0,0 +1,424 @@
+  // Create vertex with the new rung and index
+  std::pair<long unsigned int, long unsigned int> vertex(rung, idx);
+  // Store new vertex to the returned state
+  state->as<typename descartes_light::DescartesStateSpace<FloatType>::StateType>()->vertex = vertex;
```
Changed to match this now since no longer using a pair.

marrts commented:
Updated to remove all boost dependencies and rebased to emulate the new directory structure.

I think I fixed all of your minor code tidiness concerns and I addressed all of the minor change recommendations.

I made some more detailed comments and tried to address most of your concerns about how I was describing things. Hopefully it is more clear what the various functions do now.

The biggest question out there is still the issue of max_dist_ and rung_to_rung_dist_ which I tried to address in more detail in comments above.

Levi-Armstrong commented:
@marip8, @marrts, and @colin-lewis-19 what is left for this to be merged?

marrts commented:
The ompl planners crash when running the unit tests and I haven't been able to figure out why.

Levi-Armstrong commented:
What is the current status of this MR?

marrts commented:
I haven't looked at this in a while, and I'm not aware of any plans for anyone to actually use this. I think it can be closed, unless someone wants to do a little bit of clean up work to merge it in.

Levi-Armstrong commented:
> I haven't looked at this in a while, and I'm not aware of any plans for anyone to actually use this. I think it can be closed, unless someone wants to do a little bit of clean up work to merge it in.

Didn't this perform better than the BGL version?

marrts commented:
> Didn't this perform better than the BGL version?

It had some really positive results, but it can be finicky. In the application, with a really large search space, we originally had in mind for it we did not end up using it as I could never get that good of results. I think it needs some refinements. Usually it would either converge very fast to a feasible solution, or timeout trying.
